pkg/build/gobuild.go

```diff
@@ -939,3 +998,158 @@ func (pm *platformMatcher) matches(base *v1.Platform) bool {
+	if err != nil {
+		return nil, err
+	}
+
+	if buildid == "" {
+		return nil, fmt.Errorf("no buildid for %s", file)
+	}
+
+	btod, err := c.readBuildToDiff(file)
+	if err != nil {
+		return nil, err
+	}
+	dtod, err := c.readDiffToDesc(file)
+	if err != nil {
+		return nil, err
+	}
+
+	diffid, ok := btod[buildid]
+	if !ok {
+		return nil, fmt.Errorf("no diffid for %q", buildid)
+	}
+
+	desc, ok := dtod[diffid]
+	if !ok {
+		return nil, fmt.Errorf("no desc for %q", diffid)
+	}
+
+	return &lazyLayer{
+		diffid: diffid,
+		desc:   desc,
+		buildLayer: func() (v1.Layer, error) {
+			return buildLayer(appPath, file, platform)
+		},
+	}, nil
+}
+
+// Compute new layer metadata and cache it in-mem and on-disk.
+func (c *layerCache) put(ctx context.Context, file string, layer v1.Layer) error {
+	buildid, err := getBuildID(ctx, file)
+	if err != nil {
+		return err
+	}
+
+	desc, err := partial.Descriptor(layer)
+	if err != nil {
+		return err
+	}
+
+	diffid, err := layer.DiffID()
+	if err != nil {
+		return err
+	}
+
+	btod, ok := c.buildToDiff[file]
+	if !ok {
+		btod = buildIDToDiffID{}
+	}
+	btod[buildid] = diffid
+
+	dtod, ok := c.diffToDesc[file]
+	if !ok {
+		dtod = diffIDToDescriptor{}
+	}
+	dtod[diffid] = *desc
+
+	// TODO: Implement better per-file locking.
+	c.Lock()
+	defer c.Unlock()
+
+	btodf, err := os.OpenFile(filepath.Join(filepath.Dir(file), "buildid-to-diffid"), os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		return err
+	}
+	defer btodf.Close()
+
+	dtodf, err := os.OpenFile(filepath.Join(filepath.Dir(file), "diffid-to-descriptor"), os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		return err
+	}
+	defer dtodf.Close()
+
+	enc := json.NewEncoder(btodf)
+	enc.SetIndent("", "  ")
+	if err := enc.Encode(&btod); err != nil {
+		return err
+	}
+
+	enc = json.NewEncoder(dtodf)
+	enc.SetIndent("", "  ")
+	if err := enc.Encode(&dtod); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *layerCache) readDiffToDesc(file string) (diffIDToDescriptor, error) {
+	if dtod, ok := c.diffToDesc[file]; ok {
+		return dtod, nil
+	}
+
+	dtodf, err := os.Open(filepath.Join(filepath.Dir(file), "diffid-to-descriptor"))
+	if err != nil {
+		return nil, err
+	}
+	defer dtodf.Close()
+
+	var dtod diffIDToDescriptor
+	if err := json.NewDecoder(dtodf).Decode(&dtod); err != nil {
+		return nil, err
+	}
+	c.diffToDesc[file] = dtod
+	return dtod, nil
+}
+
+func (c *layerCache) readBuildToDiff(file string) (buildIDToDiffID, error) {
+	if btod, ok := c.buildToDiff[file]; ok {
+		return btod, nil
+	}
+
+	btodf, err := os.Open(filepath.Join(filepath.Dir(file), "buildid-to-diffid"))
+	if err != nil {
+		return nil, err
+	}
+	defer btodf.Close()
+
+	var btod buildIDToDiffID
+	if err := json.NewDecoder(btodf).Decode(&btod); err != nil {
+		return nil, err
+	}
+	c.buildToDiff[file] = btod
+	return btod, nil
+}
+
+func getBuildID(ctx context.Context, file string) (string, error) {
+	cmd := exec.CommandContext(ctx, "go", "tool", "buildid", file)
```

Comment:
Does this assume a minimum Go version we need to make users aware of?

Reply:
nvm, looks like this was introduced in Go 1.10, which is ancient history as far as we're concerned.